src/ML/src/MANNAutoregressiveInputBuilder.cpp

```diff
@@ -0,0 +1,269 @@
+    ok = ok
+         && getParameter("max_facing_direction_angle_side_opposite_sign",
+                         m_pimpl->maxFacingDirectionAngleSideOppositeSign);
+    ok = ok
+         && getParameter("max_facing_direction_angle_side_same_sign",
+                         m_pimpl->maxFacingDirectionAngleSideSameSign);
+    ok = ok && getParameter("number_of_knots", m_pimpl->numberOfKnots);
+    ok = ok && m_pimpl->curve.curve.initialize(ptr);
+
+    // if everything went fine we can resize the matrices
+    if (ok)
+    {
+        m_pimpl->output.desiredFutureBaseTrajectory.resize(2, m_pimpl->numberOfKnots);
+        m_pimpl->output.desiredFutureBaseVelocities.resize(2, m_pimpl->numberOfKnots);
+        m_pimpl->output.desiredFutureFacingDirections.resize(2, m_pimpl->numberOfKnots);
+
+        m_pimpl->curve.initialPoint.setZero();
+        m_pimpl->frontalDirection << 1, 0;
+
+        m_pimpl->fsm = Impl::FSM::Initialized;
+    }
+
+    return ok;
+}
+
+bool MANNAutoregressiveInputBuilder::advance()
+{
+    constexpr auto logPrefix = "[MANNAutoregressiveInputBuilder::advance]";
+    if (m_pimpl->fsm == Impl::FSM::Idle)
+    {
+        log()->error("{} Please initialize the class before calling advance.", logPrefix);
+        return false;
+    }
+
+    m_pimpl->fsm = Impl::FSM::OutputInvalid;
+
+    // evaluate the
+
+    m_pimpl->curve.controlPoint[1] = 0;
+
+    if (m_pimpl->input.motionDirection[0] >= 0)
+    {
+        m_pimpl->curve.controlPoint[0]
+            = m_pimpl->ellipsoidScalingFactor
+              * std::min(m_pimpl->ellipsoidForwardAxis, m_pimpl->input.motionDirection[0]);
+        m_pimpl->curve.finalPoint[0] = m_pimpl->input.motionDirection[0] * m_pimpl->ellipsoidScalingFactor
+                        * m_pimpl->ellipsoidForwardAxis;
+        m_pimpl->curve.finalPoint[1] = m_pimpl->input.motionDirection[1] * m_pimpl->ellipsoidScalingFactor
+                        * m_pimpl->ellipsoidSideAxis;
+    } else
+    {
+        m_pimpl->curve.controlPoint[0]
+            = m_pimpl->ellipsoidScalingFactor
+              * std::max(-m_pimpl->ellipsoidBackwardAxis, m_pimpl->input.motionDirection[0]);
+        m_pimpl->curve.finalPoint[0] = m_pimpl->input.motionDirection[0] * m_pimpl->ellipsoidScalingFactor
+                        * m_pimpl->ellipsoidBackwardAxis;
+        m_pimpl->curve.finalPoint[1] = m_pimpl->input.motionDirection[1] * m_pimpl->ellipsoidScalingFactor
+                        * m_pimpl->ellipsoidSideAxis;
+    }
+
+    m_pimpl->output.desiredFutureBaseTrajectory
+        = m_pimpl->curve.curve.evaluateCurve(m_pimpl->curve.initialPoint, m_pimpl->curve.controlPoint, m_pimpl->curve.finalPoint);
+
+    m_pimpl->output.desiredFutureBaseVelocities.rightCols(m_pimpl->numberOfKnots - 1)
+        = m_pimpl->output.desiredFutureBaseTrajectory.rightCols(m_pimpl->numberOfKnots - 1)
+          - m_pimpl->output.desiredFutureBaseTrajectory.leftCols(m_pimpl->numberOfKnots - 1);
```

Comment:
Why is this necessary?

Reply:
Here I added some comments in the code that explain the rationale behind https://github.com/ami-iit/bipedal-locomotion-framework/pull/683/commits/e335c02a26e55760544e2d69ea1b686e753d4c7d

moreover 
```cpp
    m_pimpl->output.desiredFutureBaseVelocities.rightCols(m_pimpl->numberOfKnots - 1)
        = m_pimpl->output.desiredFutureBaseTrajectory.rightCols(m_pimpl->numberOfKnots - 1)
          - m_pimpl->output.desiredFutureBaseTrajectory.leftCols(m_pimpl->numberOfKnots - 1);
```
should become
```cpp
    m_pimpl->output.desiredFutureBaseVelocities.leftCols(m_pimpl->numberOfKnots - 1)
        = m_pimpl->output.desiredFutureBaseTrajectory.rightCols(m_pimpl->numberOfKnots - 1)
          - m_pimpl->output.desiredFutureBaseTrajectory.leftCols(m_pimpl->numberOfKnots - 1);
```